### PR TITLE
Introduce e2e test for Stackdriver Metadata Agent

### PR DESCRIPTION
**Pull request by kawych**
**What this PR does / why we need it**:
Introduce e2e test for Stackdriver Metadata Agent

**Release note**:
```release-note
None
```

### CODE REVIEW

**Comment by kawych** [comment]
/assign @x13n 
/cc @bmoyles0117
/cc @igorpeshansky
Igor, Bryan - can one of you review this PR?

File: test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	}
+
+	// Create test pod with unique name.
+	framework.CreateExecPodOrFail(kubeClient, f.Namespace.Name, uniqueContainerName, func(pod *v1.Pod) {
+		pod.Spec.Containers[0].Name = uniqueContainerName
+	})
+	defer kubeClient.CoreV1().Pods(f.Namespace.Name).Delete(uniqueContainerName, &metav1.DeleteOptions{})
+
+	// Wait a short amount of time for Metadata Agent to be created and metadata to be exported
+	time.Sleep(120 * time.Second)
```
Comment (x13n) [inline test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go]: How about moving sleep amount to a const?

File: test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+		}
+		if resource.resourceType == "k8s_container" &&
+			resource.resourceLabels["container_name"] == containerName {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func parseResource(result interface{}) (*Resource, error) {
```
Comment (x13n) [inline test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go]: Two things:
- You're not handling errors in this function. What if some type casts fail or labels/type field is missing?
- "result" is hardly an appropriate name for the function argument here.

**Review by x13n — CHANGES_REQUESTED** [review]

File: test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go (hunk 2 above)
Comment (kawych) [reply test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go]: Done

File: test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+	. "github.com/onsi/ginkgo"
+	"io"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	instrumentation "k8s.io/kubernetes/test/e2e/instrumentation/common"
+	"reflect"
+)
+
+// OAuth2 scopes used by Stackdriver Metadata API.
```
Comment (x13n) [inline test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go]: This doesn't make sense for the metadataWaitTime right below.

File: test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+const (
+	metadataWaitTime = 120 * time.Second
```
Comment (x13n) [inline test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go]: Please also comment what this constant is.

File: test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+		}
+	}
+	return false, nil
+}
+
+func parseResource(resource interface{}) (*Resource, error) {
+	var labels map[string]string = map[string]string{}
+	resourceMap, ok := resource.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("Resource entry is type %s, but map expected", reflect.TypeOf(resource))
```
Comment (x13n) [inline test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go]: "Resource entry is of type %s, expected map[string]interface{} instead"

File: test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+	}
+	resourceType, ok := resourceMap["type"]
+	if !ok {
+		return nil, fmt.Errorf("Resource entry doesn't have specified type")
```
Comment (x13n) [inline test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go]: "Resource entry doesn't have a type specified"

File: test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+	}
+	resourceTypeName, ok := resourceType.(string)
+	if !ok {
+		return nil, fmt.Errorf("Resource type is type %s, but string expected", reflect.TypeOf(resourceType))
```
Comment (x13n) [inline test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go]: "Resource type is of type %s, expected string instead"

File: test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+	}
+	resourceLabels, ok := resourceMap["labels"]
+	if !ok {
+		return nil, fmt.Errorf("Resource entry doesn't have specified labels")
```
Comment (x13n) [inline test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go]: "Resource entry doesn't have any labels specified"

File: test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+	}
+	resourceLabelMap, ok := resourceLabels.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("Resource labels entry is type %s, but map expected", reflect.TypeOf(resourceLabels))
```
Comment (x13n) [inline test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go]: "Resource labels entry is of type %s, expected map[string]interface{} instead"

File: test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+	}
+	for label, val := range resourceLabelMap {
+		labels[label], ok = val.(string)
+		if !ok {
+			return nil, fmt.Errorf("Resource label %q is type %s, but string expected", label, reflect.TypeOf(val))
```
Comment (x13n) [inline test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go]: "Resource label %q is of type %s, expected string instead"

**Comment by kawych** [comment]
Comments applied with only slightly modified error messages for conciseness

File: test/e2e/instrumentation/monitoring/BUILD [hunk 11]
```diff
@@ -29,6 +30,7 @@ go_library(
         "//vendor/github.com/onsi/gomega:go_default_library",
         "//vendor/golang.org/x/oauth2/google:go_default_library",
         "//vendor/google.golang.org/api/monitoring/v3:go_default_library",
+        #"//vendor/google.golang.org/api/oauth2/v2:go_default_library",
```
Comment (x13n) [inline test/e2e/instrumentation/monitoring/BUILD]: Please either uncomment (if you're going to use it) or remove this line.

File: test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+	time.Sleep(metadataWaitTime)
+
+	resp, err := oauthClient.Get(endpoint)
+	if err != nil {
+		framework.Failf("Failed to call Stackdriver Metadata API %s", err)
+	}
+	if resp.StatusCode != 200 {
+		framework.Failf("Stackdriver Metadata API returned error status: %s", resp.Status)
+	}
+	metadataAPIResponse := new(bytes.Buffer)
```
Comment (bmoyles0117) [inline test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go]: When I saw this, `ioutil.ReadAll` came to mind. Would using that make this code simpler if we're only using the buffer to extract `.Bytes()`?

https://golang.org/pkg/io/ioutil/#ReadAll

Comment (kawych) [reply test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go]: Done

File: cluster/gce/config-test.sh [hunk 13]
```diff
@@ -145,6 +145,16 @@ ENABLE_CLUSTER_MONITORING="${KUBE_ENABLE_CLUSTER_MONITORING:-influxdb}"
 # TODO(piosz) remove this option once Metrics Server became a stable thing.
 ENABLE_METRICS_SERVER="${KUBE_ENABLE_METRICS_SERVER:-true}"
 
+# Optional: Metadata agent to setup as part of the cluster bring up:
+#   none        - No metadata agent
+#   stackdriver - Stackdriver metadata agent
+# Metadata agent is a daemon set that provides metadata of kubernetes objects
+# running on the same node for exporting metrics and logs.
+ENABLE_METADATA_AGENT="${KUBE_ENABLE_METADATA_AGENT:-none}"
```
Comment (x13n) [inline cluster/gce/config-test.sh]: If this is none in test by default, shouldn't it be enabled somewhere for testing?

Comment (kawych) [reply cluster/gce/config-test.sh]: It is enabled specifically for instrumentation tests: https://github.com/kubernetes/test-infra/blob/master/jobs/env/ci-kubernetes-e2e-gce-stackdriver.env

**Comment by x13n** [comment]
Once merged, please cherrypick it together with https://github.com/kubernetes/kubernetes/pull/59649 into 1.9.

**Comment by kawych** [comment]
/assign @gmarek  
Marek, can you please approve cluster/gce/config-test.sh and cluster/gce/config-default.sh changes?

**Pull request merged**